File: backend/src/main/java/com/collabcrm/crdt/LamportClock.java
package com.collabcrm.crdt;

/**
 * Lamport Clock — Logische Uhr für verteilte Systeme.

 */
public class LamportClock {

    private long counter;

    public LamportClock() {
        this.counter = 0;
    }

    public LamportClock(long initial) {
        this.counter = initial;
    }


    public synchronized long tick() {
        return ++counter;
    }


    public synchronized long receive(long received) {
        counter = Math.max(counter, received) + 1;
        return counter;
    }


    public synchronized long current() {
        return counter;
    }
}

File: frontend/src/crdt/LamportClock.ts
/**
 * Lamport Clock — Logische Uhr für verteilte Systeme (TypeScript-Version).
 *
 * Identische Logik wie LamportClock.java im Backend.
 * Jeder Browser-Tab hat seine eigene Instanz.
 *
 * Garantiert kausale Ordnung: Wenn Aktion A vor B passiert → timestamp(A) < timestamp(B)
 * Nutzt keine Systemzeit → funktioniert auch wenn Geräte-Uhren nicht synchron sind.
 */
export class LamportClock {
  private counter: number;

  constructor(initial = 0) {
    this.counter = initial;
  }

  /** Eigene Aktion: Counter +1 und neuen Timestamp zurückgeben */
  tick(): number {
    return ++this.counter;
  }

  /** Nachricht empfangen: Counter = max(eigener, empfangener) + 1 */
  receive(received: number): number {
    this.counter = Math.max(this.counter, received) + 1;
    return this.counter;
  }

  /** Aktuellen Counter-Wert lesen */
  current(): number {
    return this.counter;
  }
}
